### ISSUE TITLE
Improvement ideas for API design

### ISSUE THREAD

**Issue by sirthias** [planet42/Laika#22]
I am currently looking into Laika as a potential templating engine for a new project and really like the idea of being able to write my content in ReST, have it be parsed into an immutable AST that I can freely manipulate and then render it to HTML with the ability to embed everything in HTML templates.

It looks like Laika has everything I need and is pure Scala, yay!! :)

As I am digging my way through the Laika API (and sources) I have come across a few improvement ideas (with possibly more coming), which I'd like to use this ticket as a loosely structured dumping ground for, if that would work for you, Jens. I also might add comments to source lines directly (in the commits that introduced them).

One thing up front: I think the work that has gone into Laika so far is truly exceptional! I know first-hand how hard writing a proper Markdown parser is, so doing that, plus fully supporting ReST, plus templating, plus proper rewriting and rendering, is something that no one else has pulled off, to my knowledge, on the JVM. So please don't get me wrong, I think that Laika should get a lot more attention and I'll be seriously rooting for it, no doubt. My intent is to maybe contribute a little to its future success by providing a purely subjective outside perspective of a new user on small pieces of the API, as long as the project is still young and improvements are easy because breaking changes aren't that costly yet.

Ok, so here go:

The overarching theme of my initial feeling is that, IMHO, the API is not yet as clean, open and flexible as it might be, with the different Laika layers being somewhat conflated in certain areas. Also, the current style appears to favor inheritance over composition, which makes things a bit harder for me as a user in some cases.

For example:
At the basic level I'd assume that laika-core consists of mainly four parts:

1. A mostly anemic (i.e. logic-less) document AST model, which has no dependencies onto anything else
2. A parsing part, for getting stuff from all kinds of raw inputs into a document AST
3. A transformation part, providing all kinds of predefined logic and helpers for AST manipulation
4. A rendering part, for rendering a document to potentially many output formats

Without having thought about it much further I'd say that parts 2, 3 and 4 should be able to have few dependencies between each other and only depend on 1.

Since I am using Laika embedded and need to separate parsing from rendering I was first looking at this basic level and let the API guide me.

Looking at the parsing API I assume the core parsing logic to return the top-level AST node, which currently appears to be `Documents.Document`.
Why is it not a `case class`? Why does it contain things like `docNumber` and `rewriteRules`, which appear to be more related to transformation or even rendering?
In my application I'd like to augment the config of a `Document` with a few more things that are generated at runtime. If `Document` was a case class that be easy to do. However, how do I do it currently? I can create a new `Document` instance, copy over everything and patch the `config`, but I can't get to the private `docNumber` and `rewriteRules` members.
In general there is quite a bit of logic in the `Document` class that I wouldn't expect in an AST node.

Another example: `DocumentContext`
Simply "marrying" a `Document` with a `TemplateDocument` isn't that easy right now. I'd expect to have some logic that allows me to pass the root node of a document AST and the root node of a template AST to get back a full document AST. Currently it appears as if the `DocumentContext` ties together a document with its environment, incl. its template. Unfortunately it isn't a case class. The logic for retrieving the template is buried in the `lazy val template` rather than living on the outside. This makes it really hard to simply say "this is the document and this is the template". Also, I can't really reuse the template retrieval logic for my own purposes.
Wouldn't it be better to make `DocumentContext ` a case class, move most of the logic out and thereby making the API much more flexible and accessible?

`DocumentTree` appears to suffer from similar caveats.

The level underneath `Document`, the `Elements` types, appear to be very clean though, with no logic and only data. Excellent for flexible manipulation.
The level atop the `Elements` however looks like it's meant to be used in a very specific way rather than giving me a flexible toolbox of things that I can plug together myself, in the way that I need it.

Please feel free to correct me wherever I have not understood things correctly.
It might very well be that I am simply rubbing Laika the completely wrong way.

**Comment by jenshalm** [comment]
There is certainly room for improvement. I would prefer to avoid any redesign in the timeframe for 0.6 (adding PDF/XSL-FO support), as that release is 90% done and was dragging for a long time already. But after that I'm open to rethink APIs. I'll write a longer comment on the weekend, explaining why the current design turned out to be like it is. That will probably make it easier to think about alternatives.

**Comment by sirthias** [comment]
Thanks, Jens!

I'll give a bit more background on my use case as food for thought:

In my application all content lives on the classpath, so I am loading all documents, templates, etc. from classpath resources, which works nicely via the "reading from files" support that Laika already has.
However, the layout of my content with regard to paths is very application specific and I don't expect Laika to support it. I am fine with simply loading and parsing documents and templates myself.

Basically I want to

1. Load and parse all documents and templates at application start, maybe apply some rewriting and build a `Map` that I can later access by document name.
2. At runtime, fetch the doc AST from the map, enrich its config with some additional settings, maybe apply some more rewriting and finally render it to HTML.

Currently I am still undecided whether I want to apply the template early (i.e. in step 1) or late (in step 2), but Laika should make it easy to do either.

I think I can do everything already with the existing implementation, but it requires some serious hacking to bend the API to my will. For example, I can't simply tell the `DocumentContext` to use a specific `TemplateDocument` that I already have. And I can't even subclass the `DocumentContext` in order to override its `template` member (which I'd consider horribly style, but it'd work), because the constructor is private. It seems I'll have to supply the DocumentContext with a fake DocumentTree so it will find and use the template instance that I already have.

Also, it's not clear to me when exactly what kind of rewriting is applied. There appears to be a forced rewriting step right after parsing, which already tries to apply a template, even though I have only just parsed a single document. Why is parsing intermingled with rewriting?

In general, it's not clear to me how responsibilities are distributed between `Document`, `DocumentContext` and `DocumentTree`. Document depends on DocumentContext (in its rewriting methods), DocumentContext depends on Document as well as DocumentTree and DocumentTree depends on Document as well as DocumentContext.

**Comment by sirthias** [comment]
I just discovered that I can simply call `template.rewrite(DocumentContext(document))` in order to manually apply my template. So no need to supply a fake `DocumentTree`, great!

**Comment by sirthias** [comment]
If I do this

```scala
val doc = Parse as ReStructuredText.withRawContent fromFile file1
val templateDoc = ParseTemplate fromFile file2
val fullDoc = templateDoc.rewrite(DocumentContext(doc))
val html = (Render as HTML from fullDoc).toString
```

then the produced HTML has this title tag

```html
<title>List(Text(This is the title!,NoOpt))</title>
```

i.e. there is a `toString` somewhere, where it shouldn't be.
Do you have an idea of how I could fix that?

**Comment by jenshalm** [comment]
Could be a bug. There are many tests around processing templates as part of transforming an entire tree, but maybe there is an untested edge case when just using a single document. I assume the relevant part in your template is simply `<title>{{document.title}}</title>`? I'll try to reproduce in the evening.

**Comment by sirthias** [comment]
Yes, right. I'll open a ticket.

**Comment by jenshalm** [comment]
Ok, finally I have some time to write a longer response. First, many thanks for your detailed comment, it really helps set priorities for the upcoming releases, in particular as feedback is currently still fairly scarce.

One thing I was not clear about from your comment is which of the areas you consider deficient. Some comments were quite general whereas others focussed mainly on the model for `Document`, `DocumentTree` and `DocumentContext`. For the latter I fully agree. This area was only added in the 0.4 release where tree processing was introduced and subsequently grew to adjust to new features in the following releases. In contrast to many other API layers it hasn't seen a single true polishing cycle yet and this shows. 

So the good news is that your comment motivated me to finally rewrite this area. I'll do this as the first thing after the next release. If you are interested, I'll do a proper PR so you can review it first. Normally I don't bother as I simply don't have any reviewers (yet). But since this will take a while, I'll add a bit of background around why this layer is designed like this, which hopefully will make it easier to use for now (and probably should find its way into the scaladoc). 

Only `DocumentTree` and `Document` are part of the AST, they have been added on top of the original AST, which only supported single-document processing and may still feel a bit alien because of that. 

`DocumentContext` might seem a bit confusing and the poorly chosen name probably adds to that. I'm thinking of renaming it to `TreeCursor`, since that is what it essentially is. The API would be adjusted as well, to make it appear as a true cursor. It is not part of the AST and only exists during rewriting. The signature of a simple rewrite rule is `Element => Option[Element]` but this only allows to examine the target element itself (and its children). For most rules this is sufficient, but some rules need to get the full picture, e.g. those that process cross links. For that reason a second signature exists: `DocumentContext => Element => Option[Element]`. It allows to capture (read-only) access to the full tree (in its pre-rewrite state) for the rewrite rule. Rule creation is the only intended use of this class in application code.  Which means that `TemplateDocument` should get a `rewrite(Document)` method so that you do not have to create an empty context yourself.

The context/cursor is an extra layer as the AST has to be unidirectional, to allow for an immutable model with efficient creation of modified copies. Therefore a separate model has to provide a view onto sibling and parent layers of the AST. This general principles is something I would retain in the redesign, but the implementation could indeed be much cleaner and more compositional. The 3 classes can become case classes, too, but only after the cleanup, as it currently would let too much low-level ugliness surface. I definitely want to try to slim down the 12-parameter `DocumentTree` constructor. :-)

Some of the logic currently buried in these classes will definitely move elsewhere, but I don't think I'd like to see them become fully anaemic. The one thing I'd like to retain is the various `rewrite` methods. They are available on the `Element` APIs, too, and moving this to external helper classes would make it less convenient to use. 

Regarding the parser already pre-applying some rewrite rules, this is about the tricky balance between strict separation of concerns and ease of use. The default rules are applied by default (unless you use `asRawDocument`) to provide a route of least surprise. You can take a document produced by a parser and pass it directly to a renderer or examine its content without finding/rendering temporary nodes (like unresolved link references). There is a marker trait `Temporary` which some `Element` types mix in, and documents which still contain them cannot be rendered properly as no renderer knows what to do with them. 

If you run into other issues, where the API is inconvenient or non-obvious to use, it would be good to know.

**Comment by jenshalm** [comment]
Now that version 0.6 with PDF support is finally out, I will tackle the design of the Document API as the next task. Since this ticket was a rather broad discussion, I'll close this now and added a new ticket #31 for the redesign. Thanks a lot for your feedback!
